src/compiler/utilities.ts

```diff
@@ -10069,18 +10078,95 @@ export const emptyFileSystemEntries: FileSystemEntries = {
  * @internal
  */
 export function matchPatternOrExact(patternOrStrings: readonly (string | Pattern)[], candidate: string): string | Pattern | undefined {
-    const patterns: Pattern[] = [];
-    for (const patternOrString of patternOrStrings) {
-        if (patternOrString === candidate) {
-            return candidate;
+    let matchableStringSet: Set<string>;
+    let sortedPatterns: Pattern[];
+
+    const cacheEntry = patternOrStringsCache.get(patternOrStrings);
+    if (cacheEntry !== undefined) {
+        ({ matchableStringSet, sortedPatterns } = cacheEntry);
+    }
+    else {
+        matchableStringSet = new Set();
+        sortedPatterns = [];
+
+        for (const patternOrString of patternOrStrings) {
+            if (typeof patternOrString === "string") {
+                matchableStringSet.add(patternOrString);
+            }
+            else {
+                sortedPatterns.push(patternOrString);
+            }
+        }
+
+        sortedPatterns.sort((a, b) => compareStringsCaseSensitive(a.prefix, b.prefix));
+
+        patternOrStringsCache.set(patternOrStrings, {
+            matchableStringSet,
+            sortedPatterns,
+        });
+    }
+
+    if (matchableStringSet.has(candidate)) {
+        return candidate;
+    }
+    if (sortedPatterns.length === 0) {
+        return undefined;
+    }
+
+    let index = binarySearchKey(sortedPatterns, candidate, getPatternPrefix, compareStringsCaseSensitive);

```

Comment:
In `Rush` we have to efficiently match a large array of Git file paths to project prefixes as part of hashing our input states; for that we use this class which you might find useful for this application: https://github.com/microsoft/rushstack/blob/main/libraries/rush-lib/src/logic/LookupByPath.ts

Reply:
For some numbers, that tool matches 142936 unique file paths to 1279 unique path prefixes in 62 milliseconds on my codespace machine (single threaded).